### PR TITLE
add the ability to communicate with the new go based authz microservice

### PR DESCRIPTION
**Pull request by omer9564**

### CODE REVIEW

File: horizon/gopal/client.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+from loguru import logger
+from opal_client import OpalClient
+from opal_client.engine.runner import PolicyEngineRunner
+
+from horizon.config import sidecar_config
+from horizon.gopal.runner import GopalRunner
+
+
+class GOPALClient(OpalClient):
+
```
Comment (asafc) [inline horizon/gopal/client.py:13]: not sure i understand your assumptions in this class.
if you run the go service, you need to wait for it to become ready, and that should reflect in the PDP health and readiness checks. where is this code?

File: horizon/gopal/client.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        self, callback: Callable[[], Awaitable]
+    ):
+        if self.engine_runner:
+            # runs the callback after policy store is up
+            await self._run_engine_runner(callback, self.engine_runner)
+            return
+
+        # we do not run the policy store in the same container
+        # therefore we can immediately run the callback
+        await callback()
```
Comment (asafc) [inline horizon/gopal/client.py:34]: not sure i understand this, you still need this external service which is an important dependency to become ready.

File: horizon/gopal/client.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        engine_runner.register_process_initial_start_callbacks([callback] if callback else [])
+        async with engine_runner:
+            await engine_runner.wait_until_done()
+
+    async def _run_or_delay_for_engine_runner(
+        self, callback: Callable[[], Awaitable]
+    ):
+        if self.engine_runner:
+            # runs the callback after policy store is up
+            await self._run_engine_runner(callback, self.engine_runner)
```
Comment (asafc) [inline horizon/gopal/client.py:29]: are you still running OPA if GOPAL is enabled?
i thought GOPAL should now manage it.

File: horizon/enforcer/api.py [hunk 4]
```diff
@@ -268,6 +269,53 @@ async def _is_allowed(query: BaseSchema, request: Request, policy_package: str):
...
+                timeout=sidecar_config.OPA_CLIENT_QUERY_TIMEOUT,
+                raise_for_status=True,
+            ) as opa_response:
+                stats_manager.report_success()
+                return await proxy_response(opa_response)
+    except asyncio.exceptions.TimeoutError:
+        stats_manager.report_failure()
+        exc = HTTPException(
+            status.HTTP_504_GATEWAY_TIMEOUT,
+            detail="OPA request timed out (url: {url}, timeout: {timeout}s)".format(
```
Comment (asafc) [inline horizon/enforcer/api.py:294]: should be "authz service request timed out" to make the distinction from the same error when gopal is disabled.

File: horizon/enforcer/api.py [hunk 5]
```diff
@@ -268,6 +269,53 @@ async def _is_allowed(query: BaseSchema, request: Request, policy_package: str):
...
+                url=url,
+                timeout=sidecar_config.OPA_CLIENT_QUERY_TIMEOUT,
+            ),
+        )
+    except aiohttp.ClientResponseError as e:
+        stats_manager.report_failure()
+        exc = HTTPException(
+            status.HTTP_502_BAD_GATEWAY,  # 502 indicates server got an error from another server
+            detail="OPA request failed (url: {url}, status: {status}, message: {message})".format(
```
Comment (asafc) [inline horizon/enforcer/api.py:303]: same here: authz service request failed

File: horizon/enforcer/api.py [hunk 6]
```diff
@@ -268,6 +269,53 @@ async def _is_allowed(query: BaseSchema, request: Request, policy_package: str):
...
+                url=url, status=e.status, message=e.message
+            ),
+        )
+    except aiohttp.ClientError as e:
+        stats_manager.report_failure()
+        exc = HTTPException(
+            status.HTTP_502_BAD_GATEWAY,
+            detail="OPA request failed (url: {url}, error: {error}".format(
```
Comment (asafc) [inline horizon/enforcer/api.py:311]: same here: authz service request failed

**Review by asafc — CHANGES_REQUESTED** [review]
some small changes, and some questions

File: horizon/gopal/client.py (hunk 1 above)
Comment (omer9564) [reply horizon/gopal/client.py:13]: missing the reflection in the healthcheck, will add

File: horizon/gopal/client.py (hunk 3 above)
Comment (omer9564) [reply horizon/gopal/client.py:29]: no, the OPAL service manages OPA

Comment (omer9564) [reply horizon/gopal/client.py:29]: Didn't want to add all the complexity of managing the OPA service

File: horizon/gopal/client.py (hunk 2 above)
Comment (omer9564) [reply horizon/gopal/client.py:34]: I call it in the start_client_background_tasks

Comment (omer9564) [reply horizon/gopal/client.py:34]: I actually thought to do it first in the _run_or_delay_for_engine_runner function but didn't do it there eventually, so the _run_or_delay_for_engine_runner function is literally a copy paste from the parent class, will remove it

File: horizon/enforcer/api.py (hunk 4 above)
Comment (omer9564) [reply horizon/enforcer/api.py:294]: gopal calls opa, there is no way to distinguish between the two

File: horizon/enforcer/api.py (hunk 5 above)
Comment (omer9564) [reply horizon/enforcer/api.py:303]: gopal calls opa, there is no way to distinguish between the two

File: horizon/enforcer/api.py (hunk 6 above)
Comment (omer9564) [reply horizon/enforcer/api.py:311]: gopal calls opa, there is no way to distinguish between the two

File: horizon/enforcer/api.py (hunk 4 above)
Comment (asafc) [reply horizon/enforcer/api.py:294]: i still think a different error message would help us understand faster what version of the PDP the customer is running and whether or not gopal is enabled.

File: horizon/gopal/client.py (hunk 3 above)
Comment (asafc) [reply horizon/gopal/client.py]: ok for now

File: horizon/gopal/client.py (hunk 1 above)
Comment (asafc) [reply horizon/gopal/client.py]: looks like you fixed it :)

**Review by asafc — APPROVED** [review]
you still have merge conflicts, but overall looks good.
please check the healthcheck is working e2e with gopal.

**Pull request merged**